CRM-17608: Merge to DOCx or ODT template

monishdeb opened the pull request:
* [CRM-17608: Merge to DOCx or ODT template](https://issues.civicrm.org/jira/browse/CRM-17608)

monishdeb commented on CRM/Utils/PDF/Document.php:
```diff
@@ -95,4 +95,64 @@ public static function toTwip($value, $metric) {
+   * @return string
+   *   File content of document as text
+   */
+  public static function doc2Text($filePath, $type) {
```
@colemanw I didn't found any method to extract the content as a text from document file so I can use it for token replacement, so here's my solution for now as per https://github.com/PHPOffice/PHPWord/issues/490

monishdeb commented on templates/CRM/Admin/Form/MessageTemplates.tpl:
```diff
@@ -124,3 +147,41 @@
+      function toggleByFileType(type) {
+        var show = (type == 'odt' || type == 'docx') ? false : true;
+        $("#msg_html").toggle(show);
+        $("#msg_text").toggle(show);
```
Trust me I tried to join the selectors but don't knw why this leads to distorted HTML accordion.

monishdeb commented on CRM/Utils/PDF/Document.php:
```diff
@@ -95,4 +95,64 @@ public static function toTwip($value, $metric) {
+      $phpWord->save($newpath, 'HTML');
+    }
+
+    $relPath = str_replace('/Users/monish/www/civicrm-master/', CRM_Utils_System::baseURL(), $newpath);
```
@colemanw  sorry for the hardcording but is there any function that turns absolute url to server-sided url ? (didn't found one :( )

colemanw replied:
See `\Civi::paths()->getPath()` and `\Civi::paths()->getUrl`

monishdeb commented:
@colemanw done with the changes. Please have a look.

colemanw commented:
jenkins, retest this please

colemanw commented:
Hi @monishdeb I've tested out the [demo site](http://core-8474-2rco3.test-ubu1204-5.civicrm.org/) for this PR and noticed a couple things:

1. After merging a document for a single contact, I get this:
![screenshot from 2016-06-22 20 19 56](https://cloud.githubusercontent.com/assets/2874912/16288143/29f86764-38a6-11e6-9f12-110c42e2f41c.png)
2. After merging a docx for a single contact, the `{contact.first_name}` token was not replaced. I have not tried other tokens.
3. I think the selection of document types on the "message template" screen is confusing; also we don't actually care at that point what type of document is being uploaded. I think we should change it to a radio button with choices for "Compose on-screen" or "Upload document"

monishdeb commented:
Issue 1 and 2 occur when tbsZip->fileRead() is unable to read the content of docx file and return blank content. Especially when there is a token in content. However there is no problem with odt file. In my local docx with plain text only works fine but fails when there is any token. Will try to find the fix.

colemanw commented on CRM/Admin/Form/MessageTemplates.php:
```diff
@@ -169,7 +168,9 @@ public function buildQuickForm() {
     $this->add('text', 'msg_title', ts('Message Title'), CRM_Core_DAO::getAttribute('CRM_Core_DAO_MessageTemplate', 'msg_title'), TRUE);
 
-    $this->add('select', 'file_type', ts('File Type'), CRM_Core_SelectValues::documentFormat());
+    $options = array(ts(' Compose On-screen'), ts(' Upload Document'));
```
@monishdeb it's best not to put extra whitespace inside ts strings.

monishdeb replied:
Just for clarity I added a space between radio element and its label as its not possible via tpl. Anyways will trim that.

colemanw commented on CRM/Admin/Form/MessageTemplates.php:
```diff
@@ -169,7 +168,9 @@ public function buildQuickForm() {
 
-    $this->add('select', 'file_type', ts('File Type'), CRM_Core_SelectValues::documentFormat());
+    $options = array(ts(' Compose On-screen'), ts(' Upload Document'));
+    $this->addRadio('file_type', ts('I want to'), $options);
```
Maybe "Source" would be clearer than "I want to"

colemanw commented:
@monishdeb any luck tracking down that problem with tokens?

monishdeb commented:
Nope. tired with both tblzip and phpword but no success :/

monishdeb commented:
Updated the PR

colemanw commented:
@monishdeb needs rebase please.

monishdeb commented:
Done

colemanw commented:
Thanks :)

colemanw commented:
@monishdeb doing some more testing. What happens when I try to edit and resave a template containing a doc (not making any changes, just edit + save):

![screenshot from 2016-07-12 14 25 45](https://cloud.githubusercontent.com/assets/2874912/16778574/97945134-483c-11e6-8377-ad94033f8834.png)

I think the "compose on-screen" and "upload document" radio buttons should probably be frozen when editing. But the document validation needs fixing - the field should not be required if the delete button wasn't clicked.

colemanw commented:
Also @monishdeb can you please make me a collaborator in your personal civicrm-core repo so I can push changes?

colemanw commented:
@monishdeb I think I found another bug. Steps to reproduce:

1. Create a new template using an .odt file
2. Go to a contact and select print/merge document activity.
3. Select the template.

Bug 1: The file type selector and the compose on-screen options do not disappear.
Bug 2: The resulting file is blank. The following error is seen:

![screenshot from 2016-07-12 14 44 12](https://cloud.githubusercontent.com/assets/2874912/16779240/5f0afd10-483f-11e6-9fe7-0191dccd46a1.png)

colemanw commented:
I just tried a second test where I uploaded an .odt directly instead of using a template. This time the resulting document came out with tokens replaced but all formatting was lost. It appears you are using the text only version of the document instead of preserving the original?

Again, the "document format" selector is redundant when a document has already been uploaded. We do not need a feature for transforming a document from one format to another. We simply want to replace tokens in the document and otherwise not alter it or mess up the formatting.

![screenshot from 2016-07-12 14 51 12](https://cloud.githubusercontent.com/assets/2874912/16779443/407e33ac-4840-11e6-9429-401c260f5260.png)

monishdeb commented:
@colemanw please check my inline explanation under each comment. 
For the last point 
1. I am not removing any style of odt/docx content BUT extracting only the document body here https://github.com/civicrm/civicrm-core/pull/8474/files#diff-3a69921ed714582ca34cc7c0b68c9331R213 and replacing with the token replaced body keeping header, footer, style etc. intact. 
2. Ok will look into it about hiding the 'Document Type'.

monishdeb commented:
Done the additional fixes https://github.com/civicrm/civicrm-core/pull/8474/commits/c7226ef0e7c239d692f28e3af8905765d96e8a02 :
1. Freeze source radio button on messageTemplate edit form.
2. Hide document type on uploading template on screen

colemanw commented:
@monishdeb looks like you may have done a force-push which deleted my commit. I've added it back, no worries :)
In future if your push is rejected you could do a `git pull --rebase` then (non force) push.

monishdeb commented:
@colemanw oops sorry .. I accidentally did ```git pull --rebase upstream master```.

colemanw commented:
That's OK. `git cherry-pick` makes it easy to add back commits that were accidentally deleted.

colemanw commented:
@monishdeb I think we should also hide the "Page format" accordion when selecting a file, since those controls will not affect anything.

More seriously, I noticed that when selecting a *template* which contains a file, none of the fields are hidden that would normally be when uploading a file directly.
IMO the proper behavior when selecting a file-based template would be to hide everything else (everything) and display some text saying something like "$file_name will be merged."

colemanw commented:
@monishdeb I fixed the missing formatting by outputting the document directly instead of going through phpWord. And I worked around the problem of missing tokens in docx files also.
Can you fix the cosmetic stuff I mentioned yesterday and this project will be in good shape!

monishdeb commented:
@colemanw thanks a lot for you help.  
Please check my comment above on the second UI changes you mentioned. That document body act as a preview of the chosen template document.

colemanw commented:
Ok that is the problem. CiviCRM security policy is no direct access to that directory, so you need a php callback to access those files.

monishdeb commented:
@colemanw I have made the style changes that include 
1. Changing title of 'Document Body' to 'Preview Document' via js
2. Hiding pdf page format on selection of document template or uploading document

However regarding the access issue I later found that phpWord is already using ```fwrite($filepath, $content)``` (see https://github.com/PHPOffice/PHPWord/blob/develop/src/PhpWord/Writer/AbstractWriter.php#L335) so it is no longer a issue in writing generated HTML file in upload directory. This is the code which is responsible to generate HTML version of chosen document file and save it to upload directory https://github.com/civicrm/civicrm-core/pull/8474/files#diff-3a69921ed714582ca34cc7c0b68c9331R155.  I also have found instances where we are using file_put_content($filepath, $content) php function to write file in restricted upload directory.

colemanw commented on templates/CRM/Mailing/Form/InsertTokens.tpl:
```diff
@@ -141,8 +141,10 @@ function selectValue( val, prefix) {
-    cj('.crm-html_email-accordion').toggle(hide);
+    cj('.crm-html_email-accordion, .crm-pdf-format-accordion').toggle(hide);
     cj('.crm-document-accordion').toggle(!hide);
+    cj('.crm-document-accordion .crm-accordion-header label').text(ts('Preview Document'));
```
This is a `.tpl` file so needs to use `{/literal}{ts}` instead of `ts()`.

Also shouldn't the label get changed to 1 of 2 possible strings conditionally based on the variable `hide`?

colemanw commented:
@monishdeb thanks for the style fix.

I'm not sure if we are understanding each other about the preview file. There is no problem with writing files to the upload directory. In my sandbox they are appearing just fine. The problem is with *reading* them. The `.htaccess` prevents this, and that's by design. The upload directory is *not* browsable, as it may contain sensitive documents. Indeed the documents we are dealing with in this issue may be sensitive as well, so we really should not be placing them in any directory that is browsable to the world.

Another related problem is that these `.html` files stick around cluttering up the upload directory forever.

I think the solution to both of these problems would be a page callback which properly enforces permissions and generates the file content on-the-fly without storing it as a file anywhere.

The last bug that needs dealing with is that no html is getting generated for docx files and we are still getting notices when merging docx files. Can you try to debug this?

> Notice: Trying to get property of non-object in PhpOffice\PhpWord\Reader\ODText\Content->read() (line 42 of sites/all/modules/civicrm/vendor/phpoffice/phpword/src/PhpWord/Reader/ODText/Content.php).
> Notice: Trying to get property of non-object in PhpOffice\PhpWord\Shared\XMLReader->getElement() (line 116 of sites/all/modules/civicrm/vendor/phpoffice/phpword/src/PhpWord/Shared/XMLReader.php).

monishdeb commented:
I have made the ts change as I later realised i have already allotted a separate accordion for preview so we no longer need to change the label via js rather then renaming the header label https://github.com/civicrm/civicrm-core/pull/8474/files#diff-b5a9013b9227f242b92a47431fc45240R97 .

On the otherhand, reason why I stored it in upload dir because iframe got a limitation as it only accept filepath in 'src' and which also needs to be readable, and I did't find any other way to show preview, by extracting HTML content to iframe(like in WYSIWYG). I thought since we are already using upload/ for temporary storage like to store import file so why not we should use it to store genrated HTML and that will be deleted for sure after task is done programmatically.

monishdeb commented:
Lemme try again to do it via page callback

monishdeb commented:
Yup I got an answer :) updating PR in a hour.

monishdeb commented:
@colemanw this is my change https://github.com/civicrm/civicrm-core/pull/8474/commits/9cbcd63f8150d0bd0fe82a2c371b9fe9b8cfed84 :

1. Now we are retrieving the the HTML content of the corresponding document file rather then generating and using the saved HTML file in upload/ unlike earlier. 
2. Few style change and optimisation.

colemanw commented on templates/CRM/Mailing/Form/InsertTokens.tpl:
```diff
@@ -140,7 +140,7 @@ function selectValue( val, prefix) {
 
   cj.post( dataUrl, {tid: val}, function( data ) {
-    var hide = (data.document_path && isPDF) ? false : true;
+    var hide = (data.document_body && isPDF) ? false : true;
```
Why did you change this?

monishdeb replied:
Bcoz I renamed it to document_body here https://github.com/civicrm/civicrm-core/pull/8474/files#diff-ed1056b02289bcdce4d67408f8e21253R60

monishdeb commented:
@colemanw I have added unit test, changed docx page break which is now working fine (didn't found proper page-break for odt) and some changes in core class due to unit test. Please have a look https://github.com/civicrm/civicrm-core/pull/8474/commits/b87ed98a71b5941bb38835aa6c176852fc0d4573

monishdeb commented:
Jenkin test this please

monishdeb commented:
Jenkin test this please

colemanw commented:
This all looks good @monishdeb.

What's the issue with .odt page breaks? Did you try the same technique of examining a sample document with page breaks in it?

monishdeb commented:
@colemanw yes I tried the same technique but the page-break pattern of odt aren't same and complex. Picked up all possible pattern but didn't helped much :/

colemanw commented:
Jenkins, retest this please

colemanw commented:
@monishdeb I think we should merge this because it's working great except for that one minor issue.

One idea I have for fixing the odt page break issue: is there a pattern you could search for in the existing document using regex?
